Navigation Link block: add RichText split-at-end/merge/remove behavior

ZebulanStanphill opened the pull request:
## Description
Closes #21743.
Closes #18208.

This PR adds several RichText behavior enhancements to Navigation Link blocks. It introduces the following behaviors:
- Removing an empty Navigation Link using <kbd>Backspace</kbd> or <kbd>Delete</kbd>.
- Creating a new Navigation Link by pressing <kbd>Enter</kbd> at the end of an existing one.
- Merging two adjacent Navigation Links by pressing <kbd>Backspace</kbd> or <kbd>Delete</kbd>. The labels are merged; all other attributes are kept as they are on the first block while those on the second are discarded. Children of the first block are kept, but children of the second block are lost. (No idea how to change this; ideally, I would want all the children to be moved under the first block.)

Notably, the ability to split an existing Navigation Link into two has been omitted, since this would cause a loss of all its children. I have no idea how to add this behavior and preserve children; I suspect it may not be possible to fix without improvements to the RichText APIs.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ZebulanStanphill commented:
I thought this would be a simple fix, but I've run into an issue I didn't expect. After pressing backspace and removing the block, focus doesn't move to the previous Navigation Link like you would expect. No idea why. :confused:

talldan commented:
@ZebulanStanphill From what I can gather, `onMerge` might have to be specified at well. I had a look at how the `replaceBlocks` action gets called when deleting a paragraph or heading, and it seems like it's called from a dispatch of `MERGE_BLOCKS`.

ZebulanStanphill commented:
@talldan Thanks; I added merge functionality and now the cursor is moving to the previous Navigation Link (and also you can merge the text of two links now).

I'm not done yet, though. Removing/merging causes error in the JS console which reads:

>RichText needs an identifier prop that is the block attribute key of the attribute it controls. Its type is expected to be a string, but was number

![image](https://user-images.githubusercontent.com/19592990/80831801-17069100-8bb1-11ea-9209-b926573694cd.png)

I'm not really sure what this means or how to fix it.

Also, the cursor is indeed moved to the previous block, but it's put at the beginning of the text, rather than the end.

Finally, if a Navigation Link has children, the children are deleted when merging into the previous Navigation Link. Is there any way to transfer them to the other block?

talldan commented:
> I'm not done yet, though. Removing/merging causes error in the JS console which reads:

This seems to be an implementation detail of the annotations system, which I'm not sure I've ever really looked at, so I don't have more information unfortunately! Looking at other blocks like `heading`, the identifier prop seems to be defined as a string ("content", which is the same as the value attribute name, so maybe try "label".

> Also, the cursor is indeed moved to the previous block, but it's put at the beginning of the text, rather than the end.

Yeah, that's curious. I'll try to have a look into it at some point

> Finally, if a Navigation Link has children, the children are deleted when merging into the previous Navigation Link. Is there any way to transfer them to the other block?

I suppose it's a new behavior to have a block that can be deleted in this way that might have inner blocks. I think it's ok to try out like it is initially, but we could maybe flag it for design feedback, either on this PR or as a separate issue.

ZebulanStanphill commented:
> This seems to be an implementation detail of the annotations system, which I'm not sure I've ever really looked at, so I don't have more information unfortunately! Looking at other blocks like `heading`, the identifier prop seems to be defined as a string (e.g. "content", which is the same as the value attribute name, so maybe try "label".)

So how do I add the identifier prop? I can't seem to find any documentation about it.

ZebulanStanphill commented:
I had hoped that a rebase and adjustment after #22292 would fix some of the issues in this branch, but I'm still running into all of the same issues as before.

ZebulanStanphill commented:
Well, it turns out that all I had to do to fix the JS console errors was literally just add `identifier="label"` to the `RichText`. For some reason it never occurred to me that the identifier prop was literally just called "indentifier", and I somehow missed it when looking at other blocks.

Merging/removing is now working as expected.

There is still one minor issue: removing a Navigation Link nested inside another Navigation Link doesn't focus the RichText of the parent like I would expect. Actually, it's not really clear where exactly the focus is. You have to tab or shift tab to get back to the standard tabbable elements. This also happens when backspace-removing a text block in a Group, so the issue isn't isolated to the Navigation block; because of that, I don't think it's a blocker for this issue.

I think this PR is now ready for review.

ZebulanStanphill commented:
I updated this PR, using a similar approach as #22436. It's now working much better. I've updated the PR title and description to better describe all the behavior this PR adds.

talldan commented on packages/block-library/src/navigation-link/index.js:
```diff
@@ -12,17 +12,20 @@ import edit from './edit';
-
 	__experimentalLabel: ( { label } ) => label,
-
+	merge( a, { label: bLabel = '' } ) {
+		return {
+			...a,
+			label: a.label + bLabel,
+		};
+	},
```
I see the variable names are used in the other PR, but it's against the coding standards and it should probably have been caught during review of that:
https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#naming-conventions

> Variable and function names should be full words, using camel case with a lowercase first letter.

I'm guessing here that the object is an `attributes` object.

I've seen `left` and `right` (e.g. `leftAttributes`, `rightAttributes`) used to distinguish before with similar operations, like when sorting, which I think makes sense here as there's a left block and a right block being merged.

ZebulanStanphill replied:
Thanks; I've revised the variable names.

talldan reviewed (changes requested):
Thanks for persisting on this @ZebulanStanphill! Looks like the PR now also fixes #18208, so I've updated the description.

If we can fix the naming I'd be very happy to approve.

ZebulanStanphill committed:
Add RichText split/merge/remove to Navigation Links.

talldan reviewed (approved):
🎉 Thanks @ZebulanStanphill!
